mm2src/common/mm_metrics/adapt.rs

```diff
@@ -0,0 +1,341 @@
+                    tags: labels_to_tags(metrics_core::Key::from_name(name.clone()).labels()),
+                    message: format!("{} {}", name, value),
+                });
+            }
+        }
+
+        for (name, mut by_labels) in gauges.drain() {
+            for (_labels, value) in by_labels.drain() {
+                output.push(PreparedMetric {
+                    tags: labels_to_tags(metrics_core::Key::from_name(name.clone()).labels()),
+                    message: format!("{} {}", name, value),
+                });
+            }
+        }
+
+        output
+    }
+
+    fn collect_json(&self) -> Result<MetricsJson, String> {
+        let mut output: Vec<MetricType> = vec![];
+
+        for (key, gauge) in self.registry.get_counter_handles() {
+            let value = gauge.get_inner().load(Ordering::Acquire);
+            let (_name, labels) = key_to_parts(&key, None);
+            output.push(MetricType::Counter {
+                key: key.to_string(),
+                labels,
+                value,
+            });
+        }
+
+        for (key, gauge) in self.registry.get_gauge_handles() {
+            let value = gauge.get_inner().load(Ordering::Acquire);
+            let (_name, labels) = key_to_parts(&key, None);
+            output.push(MetricType::Gauge {
+                key: key.to_string(),
+                labels,
+                value: value.try_into().unwrap(),
+            });
+        }
+
+        for (key, histogram) in self.registry.get_histogram_handles() {
+            let (_name, labels) = key_to_parts(&key, None);
+            let mut q_values = vec![];
+            let mut count = 0;
+            histogram.get_inner().clear_with(|values| {
+                count += values.len();
+                for value in values {
+                    q_values.push(Quantile::new(*value));
+                }
+            });
+
+            // Use default significant figures value.
+            // For more info on `sigfig` see the Historgam::new_with_bounds().
+            let sigfig = 3;
+            let histogram = Histogram::new(sigfig).unwrap();
+            let mut quantiles = hist_at_quantiles(histogram, &q_values);
+            // add total quantiles number
+            quantiles.insert("count".into(), count as u64);
+            output.push(MetricType::Histogram {
+                key: key.to_string(),
+                labels,
+                quantiles,
+            });
+        }
+
+        Ok(MetricsJson { metrics: output })
+    }
+}
+
+#[derive(Clone)]
+pub struct MmRecorder {
+    inner: Arc<Inner>,
+}
+
+impl MmRecorder {
+    pub fn handle(&self) -> MmHandle {
+        MmHandle {
+            metrics: self.inner.clone(),
+        }
+    }
+    /// Install this recorder as the global default recorder.
+    pub fn install(self) -> Result<MmHandle, String> {
+        let handle = self.handle();
+        metrics::set_boxed_recorder(Box::new(self)).map_err(|e| e.to_string())?;
+        Ok(handle)
+    }
```

Comment:
We shouldn't install the metrics using `metrics::set_boxed_recorder`.
As you can see, currently, every macro takes `metrics` argument to put/update the metric directly into the `Metrics::receiver`.
https://github.com/KomodoPlatform/atomicDEX-API/blob/039605956aff256a72dcaa570e83f7c45da74b67/mm2src/common/mm_metrics/native.rs#L19-L24

Please check if we can use [metrics::Recorder::register_count/gauge/histogram](https://docs.rs/metrics/latest/metrics/trait.Recorder.html) instead.

Reply:
yes, I'm aware but it seems we don't need the macros anymore as `metrics` itself now have a macro crates for this..i.e

`counter!("sam", 3, "type" => "maker")`
 `gauge!("got", 3.4, "type" => "taker")`
 
 etc... with lots of other functionalities [https://crates.io/crates/metrics-macros]

Reply:
As I said above, we shouldn't set the recorder to the global scope (i.e. us `metrics::set_boxed_recorder`).

But these macros require for us to set it globally.
> but it seems we don't need the macros anymore as metrics itself now have a macro crates for this..e.g
counter!("sam", 3, "type" => "maker")
gauge!("got", 3.4, "type" => "taker")

Reply:
Ok...I checked and I don't think It's possible to use these methods for installing recorder but I found something which is `clear_recorder` used for "Clears the currently configured recorder"

Reply:
As you can see, [increment_counter](https://github.com/metrics-rs/metrics/blob/metrics-v0.18.1/metrics-macros/src/lib.rs#L174) has the [following implementation](https://github.com/metrics-rs/metrics/blob/5f720a0a8f53aaee4bfa9de1d9fc470f1ed6b3c3/metrics-macros/src/lib.rs#L251-L255):
```rust
// Only do this work if there's a recorder installed.
if let Some(recorder) = metrics::try_recorder() {
    #locals
    let handle = recorder.#register_ident(#metric_key);
    handle.#op_ident(#op_value);
}
```

I think we can do right the same in our `mm_counter` macro:
```rust
#[macro_export]
macro_rules! mm_counter {
    /// Where `metrics` is expected to be either `MetricsArc` or `MetricsWeak`.
    /// `MmRecorder` should implement `metrics::Recorder` trait
    ($metrics:expr, $name:expr, $value:expr) => {{
        if let Some(mut recorder) = $crate::mm_metrics::TryRecorder::try_recorder(&$metrics) {
            let counter = recorder.register_counter(...);
            counter.increment_or_set_or_do_whatever_needed(...);
        }
    }};
    ($metrics:expr, $name:expr, $value:expr, $($label_key:expr => $label_val:expr),+) => // TODO
}
```

Reply:
Got it